### ISSUE TITLE
Dispatch Guard

### ISSUE THREAD

**Issue by Kethku** [KillTheMule/nvim-rs#10]
I'm in the process of porting neovide over to nvim-rs from neovim_lib, but I use a feature in neovim_lib which lets me get a join handle from the running neovim session. This lets me start a thread whos purpose is to stay parked until the join handle completes and then quit the process gracefully.

I'm wondering what the equivalent strategy is for nvim-rs

**Comment by KillTheMule** [comment]
I'll write up some details, you might know all this, but that's actually one of the "correctness things" that I'd like nvim-rs to get right, so it makes sense to me to make an inventory of things happening :) I'll write a TL;DR at the end, be sure to read that if nothing else :)

### `neovim-lib`

The `dispatch_guard` is a `join_handle` for the ["io thread"](https://github.com/daa84/neovim-lib/blob/293b26fb949f9ee9ed08e49415df4b4bfe51f819/src/rpc/client.rs#L233), namely the thread that reads messages from the reader and calls the handler functions on it. The `join_handle` just waits for it to finish, returning a `Result` depending on panic-or-not.

The io thread finished "regularly" under those conditions:

-  [Reading a message](https://github.com/daa84/neovim-lib/blob/293b26fb949f9ee9ed08e49415df4b4bfe51f819/src/rpc/client.rs#L236) failed
  - [Error from the reader](https://github.com/daa84/neovim-lib/blob/master/src/rpc/model.rs#L67)
  - The received message isn't well formed (wrong message type, some wrong type in the message). Notably though, this doesn't include an argument list of the wrong length.

The io thread finishes with an `Err`, i.e. the code paniced when:

- A handler function paniced (code from the user of the library)
- Decoding a message paniced (wrong argument length in a message, panic from reading (not sure this can happen)
- Writing a response failed (`encode` returns errors, but they're not handled)
- Internal panic with queue handling (this has a funky edge case in async... here, what could happen is that neovim sends a response with a wrong msgid, then this code panics)

There are some edge cases (out of memory, panic while trying to log) that I don't think are important.

### `nvim-rs`

The equivalent thing in `nvim-rs` is the [`io_future`](https://github.com/KillTheMule/nvim-rs/blob/master/src/neovim.rs#L76). You'll need to run it from your runtime, and the runtime hopefully gives you back a handle to it. This is done for you when using one of the [`create::new_*`](https://github.com/KillTheMule/nvim-rs/blob/master/src/create.rs#L103) functions. Note though that the `io_future` is independent of a runtime, while the `create::new_*` functions are dependent on tokio for now. I'll investigate how to ease using another runtime than tokio soon, but for your usecase (connecting to a child neovim) there's no way around depending on tokio, because nothing else seems to provide an asynchronous ChildStdin/ChildStdout).

So you can [`await`](https://docs.rs/tokio/0.2.9/tokio/task/struct.JoinHandle.html#impl-Future) the `io_handle` just as you'd `join` the dispatch_guard before, and act on this. You will get a [`LoopError`](https://github.com/KillTheMule/nvim-rs/blob/master/src/error.rs#L327) if something went wrong. The task is designed to never panic, but I can't guarantee that under all conditions, namely:

- Edge case from above
- A user-supplied handler function panicing
- Panic from the reader/writer (again, not sure this can even happen, the functions do return io errors)

Note that an [`error writing a response`](https://github.com/KillTheMule/nvim-rs/blob/master/src/neovim.rs#L215) will just result in logging and going on, because it's vital that that part of the code is not waited upon in the loop, so we can't very well handle errors. If your usecase needs this, please let me know, there _are_ ways (with tradeoffs to be made) to handle it if really needed.

Note also that the role of panicking is different in async, since runtimes might be single-threaded so a panic could bring down your whole process.

### TL;DR

Use the [`io_handle`](https://github.com/KillTheMule/nvim-rs/blob/master/src/create.rs#L108) and await that. There are examples you might not have seen, you can see the `io_handle` being used [here](https://github.com/KillTheMule/nvim-rs/blob/master/examples/scorched_earth.rs#L111).

*IMPORTANT*: I'm in the process of porting nvim-rs to use futures-rs rather than tokio as far as possible, to allow for different runtimes (this was a request of the author of [gnvim](https://github.com/vhakulinen/gnvim/) who want to use the runtime provided by gtk, maybe that's of interest to you, too). Please have a look at it, I'll probably merge it this weekend, and it's a breaking change (but I'm of course very interested in opinions about it).

**Comment by Kethku** [comment]
This was very enlightening
